source/tutorials/module-system/module-system.md

```diff
@@ -0,0 +1,1027 @@
+ }
+```
+
+Let's try to evaluate this with this file:
+
+```nix
+# eval.nix
+(import <nixpkgs/lib>).evalModules {
+  modules = [
+    ./default.nix
+  ];
+}
+```
+
+Then we run
+```bash
+nix-instantiate --eval eval.nix -A config.generate.script
+```
+
+Trying to evaluate the `generate.script` option however, we get an error
+that the option is used but not defined, indicating that we need to
+actually give a value to the option.
+
+
+## Type checking: Assigning integer to generate.script
+
+We can try to assign an integer to our option of type `lines`, but the
+module system correctly throws an error saying that our definition
+doesn't match the options type.
+
+```diff
+diff --git a/default.nix b/default.nix
+index cb423a9..0a9162e 100644
+--- a/default.nix
++++ b/default.nix
+@@ -5,5 +5,9 @@
+       type = lib.types.lines;
+     };
+   };
++
++  config = {
++    generate.script = 42;
++  };
+   
+ }
+```
+
+## Successful evaluation: Assigning a string to generate.script
+
+We can make type checking pass by assigning a string to the option,
+giving us our first successful evaluation of the `generate.script`
+option.
+
+In this case, we assign a script which calls the Google Maps Static API
+to generate a world map, then displaying the result using icat
+(image-cat), both of which are helper scripts.
+
+```diff
+diff --git a/default.nix b/default.nix
+index 0a9162e..24f9c34 100644
+--- a/default.nix
++++ b/default.nix
+@@ -7,7 +7,9 @@
+   };
+ 
+   config = {
+-    generate.script = 42;
++    generate.script = ''
++      map size=640x640 scale=2 | icat
++    '';
+   };
+   
+ }
+```
+
+TODO: Create derivations to get these commands
+
+## A new list option: Declaring generate.requestParams
+
+Let's introduce another option, generate.requestParams. For its type,
+we'll use `listOf <nestedType>`, which is a generic list type where each
+element has to match the given nested type. In our case we want `str` to
+be the nested type, which is a generic string type.
+
+Note that the difference between `str` and `lines` is in their merging
+behavior:
+- For `lines`, multiple definitions get merged by concatenation with
+  newlines
+- For `str`, multiple definitions are not allowed. Which in this case is
+  mostly irrelevant however, since we can't really define a list element
+  multiple times.
+
+```diff
+diff --git a/default.nix b/default.nix
+index 24f9c34..fd5027a 100644
+--- a/default.nix
++++ b/default.nix
+@@ -4,12 +4,21 @@
+     generate.script = lib.mkOption {
+       type = lib.types.lines;
+     };
++
++    generate.requestParams = lib.mkOption {
++      type = lib.types.listOf lib.types.str;
++    };
+   };
+ 
+   config = {
+     generate.script = ''
+       map size=640x640 scale=2 | icat
+     '';
++
++    generate.requestParams = [
++      "size=640x640"
++      "scale=2"
++    ];
+   };
+   
+ }
+```
+
+## Dependencies between options: Using generate.requestParams
+
+A collection of modules generally only has a single option that is meant
```

Comment:
This could be clearer. What I think you trying to say is that a user may only specify a single option, which causes several other options to then be evaluated (or something like that). When I read "a collection of modules generally only has a single option that is meant to be evaluated" I wonder how that can be the case. How can several modules all work together to provide a _single_ option? What's the point of the modules that don't contain that option? These are all questions a new user will have after reading this.